src/main/java/org/opensearch/neuralsearch/search/collector/HybridCollapsingTopDocsCollector.java

```diff
@@ -0,0 +1,316 @@
+    private ConcurrentHashMap<T, LeafFieldComparator[]> comparatorsMap;
+    private ConcurrentHashMap<T, FieldComparator<?>> firstComparatorMap;
+    private ConcurrentHashMap<T, Integer> reverseMulMap;
+    private ConcurrentHashMap<T, boolean[]> queueFullMap;
+    private final int numHits;
+    @Setter
+    TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
+    private HitsThresholdChecker hitsThresholdChecker;
+
+    HybridCollapsingTopDocsCollector(
+        GroupSelector<T> groupSelector,
+        String collapseField,
+        Sort groupSort,
+        int topNGroups,
+        HitsThresholdChecker hitsThresholdChecker
+    ) {
+        this.groupSelector = groupSelector;
+        this.collapseField = collapseField;
+        this.sort = groupSort;
+        if (topNGroups < 1) {
+            throw new IllegalArgumentException("topNGroups must be >= 1 (got " + topNGroups + ")");
+        } else {
+            this.needsScores = groupSort.needsScores();
+            SortField[] sortFields = groupSort.getSort();
+            this.reversed = new int[sortFields.length];
+
+            for (int i = 0; i < sortFields.length; ++i) {
+                SortField sortField = sortFields[i];
+                this.reversed[i] = sortField.getReverse() ? -1 : 1;
+            }
+
+            this.groupQueueMap = new ConcurrentHashMap<>();
+            this.collectedHitsPerSubQueryMap = new ConcurrentHashMap<>();
+            this.fieldValueLeafTrackersMap = new ConcurrentHashMap<>();
+            this.comparatorsMap = new ConcurrentHashMap<>();
+            this.firstComparatorMap = new ConcurrentHashMap<>();
+            this.reverseMulMap = new ConcurrentHashMap<>();
+            this.queueFullMap = new ConcurrentHashMap<>();
+
+            this.numHits = topNGroups;
+            this.hitsThresholdChecker = hitsThresholdChecker;
+        }
+    }
+
+    public static HybridCollapsingTopDocsCollector<?> createKeyword(
+        String collapseField,
+        MappedFieldType fieldType,
+        Sort sort,
+        int topNGroups,
+        HitsThresholdChecker hitsThresholdChecker
+    ) {
+        return new HybridCollapsingTopDocsCollector<>(
+            new CollapseDocSourceGroupSelector.Keyword(fieldType),
+            collapseField,
+            sort,
+            topNGroups,
+            hitsThresholdChecker
+        );
+    }
+
+    public static HybridCollapsingTopDocsCollector<?> createNumeric(
+        String collapseField,
+        MappedFieldType fieldType,
+        Sort sort,
+        int topNGroups,
+        HitsThresholdChecker hitsThresholdChecker
+    ) {
+        return new HybridCollapsingTopDocsCollector<>(
+            new CollapseDocSourceGroupSelector.Numeric(fieldType),
+            collapseField,
+            sort,
+            topNGroups,
+            hitsThresholdChecker
+        );
+    }
+
+    @Override
+    public List<CollapseTopFieldDocs> topDocs() throws IOException {
```

Comment:
Add comments in the code. Please explain what this method is doing by adding High level comment on major operations.